Sources/WhoopDIKit/Module/AccumulationDefinition.swift

```diff
@@ -0,0 +1,121 @@
+import Foundation
+
+private func wrapAccumulationProvider<Key: AccumulationKey>(name: String?, accumulationKey: Key.Type, valueProvider: @escaping (Any?) throws -> Key.AccumulatedValue) -> (Any?, Container) throws -> Any {
+    return { (params: Any?, container: Container) throws -> Any in
+        // Accumulate from the parent value if it exists, then add the current one
+        let previousValueFromContainer: Key.FinalValue?
+        do {
+            if let parent = container.parent {
+                let value: Key.FinalValue = try parent.get(name, params)
+                previousValueFromContainer = value
+            } else {
+                previousValueFromContainer = nil
+            }
+        } catch DependencyError.missingDependency {
+            previousValueFromContainer = nil // If there is no dependency, make it nil. Fail on other errors
+        }
+        return accumulationKey.accumulate(current: previousValueFromContainer ?? accumulationKey.defaultValue,
+                                          next: try valueProvider(params))
+    }
+}
+
+private func accumulate<Key: AccumulationKey>(key: Key.Type,
+                                              definition: DependencyDefinition,
+                                              accumulatedValueGetter: @escaping (Any?) throws -> Key.AccumulatedValue,
+                                              serviceDictionary: ServiceDictionary<DependencyDefinition>) {
+    if let value = serviceDictionary[definition.serviceKey] {
+        serviceDictionary[definition.serviceKey] = FactoryDefinition(name: definition.serviceKey.name, factory: { params, container in
+            guard let currentValue = try value.get(params: params, container: container) as? Key.FinalValue else {
+                fatalError() // This basically shouldn't be possible since we have two things with the same service key, but totally different value types
+            }
+            return try key.accumulate(current: currentValue, next: accumulatedValueGetter(params))
+        })
+    } else {
+        serviceDictionary[definition.serviceKey] = definition
+    }
+}
```

Comment:
The fact that we have seemingly two decorator closures here is kind of confusing. What is the role of wrapAccumulationProvider vs accumulate?